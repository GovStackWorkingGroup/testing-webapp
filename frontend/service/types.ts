export type BuildingBlockType = {
  id: string;
  buildingBlock: string; // Building block label
  timestamp: number; // Execution time from test [seconds]
  saveTime: number; // Save time in db [miliseconds]
  testsPassed: number;
  testsFailed: number;
  compatibility: number;
};

export type ProductsType = {
  _id: {
    testApp: string; // Label of the product
    testSuite: string; // Test suite label
    sourceBranch: string; // Branch on which tests were executed
    version: string; // Version on which tests were executed
  };
  compatibilities: BuildingBlockType[];
  overallCompatibility: number;
  lastUpdate: number;
};

export type BuildingBlockEndpointTest = {
  uri: string;
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | '';
  endpoint: string;
  passed: boolean;
  details: TestsDetailsType[];
};

export type TestsDetailsType = {
  scenario: string;
  steps: {
    text: string;
    result: boolean;
    type: string;
  }[];
};

export type ProductsListType = ProductsType[];

export type BuildingBlockTestSummary = {
  compatibilities: BuildingBlockType;
  data: BuildingBlockEndpointTest[];
  count: number;
};

export type SingleComplianceItem = {
  _id: string;
  bb: string;
  bbVersion: string;
  deploymentCompliance: number;
  interfaceCompliance: number;
  requirementSpecificationCompliance: number;
  softwareName: string;
  softwareVersion: string;
  status: number;
  submissionDate: string;
};

export type ComplianceList = {
  count: number;
  data: Record<string, SingleComplianceItem[]>;
};

// Software compliance details
export type ComplianceDetails = {
  level: number;
  note?: string;
};

export type ComplianceItem = {
  requirements: ComplianceDetails;
  interface: ComplianceDetails;
  bbVersion: string;
  deploymentCompliance: { requirement: string; level: number }[];
};

export type Compliance = {
  bbName: string;
  bbVersions: ComplianceItem[];
};

export type SoftwareDetailsType = {
  _id: string;
  logo: string;
  website: string;
  documentation: string;
  pointOfContact: string;
  compliance: [
    {
      softwareVersion: string;
      bbDetails: Compliance[];
    }
  ];
  softwareName: string;
}[];

export type BBDetailsType = {
  [key: string]: {
    requirementSpecificationCompliance: {
      crossCuttingRequirements: RequirementsType[];
      functionalRequirements: RequirementsType[];
    };
    interfaceCompliance: {
      testHarnessResult: string;
      requirements: RequirementsType[];
    };
    deploymentCompliance: number;
  };
};

export type SoftwareDraftDetailsType = {
  deploymentCompliance: DeploymentCompliance;
  description: string;
  documentation: string;
  email: string;
  expirationDate: string;
  formDetails: {
    version: string;
    bbDetails: BBDetailsType;
  }[];
  logo: string;
  softwareName: string;
  status: number;
  uniqueId: string;
  website: string;
  version: string;
};

export type Requirement = {
  requirement: string;
  comment: string;
  fulfillment: number;
  status: number;
};
export type DeploymentCompliance = {
  documentation: string | File;
  deploymentInstructions: string | File;
  requirements: Requirement[];
};

export type ComplianceVersion = {
  version: string;
  bbDetails: Map<string, SingleComplianceItem>;
};

export type SoftwareDraftToUpdateType = {
  softwareName?: string;
  logo?: File;
  website?: string;
  documentation?: string;
  pointOfContact?: string;
  compliance?: ComplianceVersion[];
  uniqueId?: string;
  expirationDate?: Date;
  deploymentCompliance?: Partial<DeploymentCompliance>;
};

export type IRSCSoftwareDraftToUpdateType = {
  bbSpecification: string;
  bbVersion: string;
  dateOfSave: string;
  requirements: {
    crossCutting: RequirementsType[];
    functional: RequirementsType[];
    interface: RequirementsType[];
  };
  interfaceCompliance: {
    testHarnessResult: string;
    requirements: [];
  };
};

export type ComplianceRequirementsType = {
  bbName: string;
  bbKey: string;
  bbVersion: string;
  dateOfSave: string;
  requirements: {
    crossCutting: RequirementsType[] | never[];
    functional: RequirementsType[] | never[];
    interface: RequirementsType[] | never[];
  };
  interfaceCompliance: {
    testHarnessResult: string;
    requirements: RequirementsType[] | [];
  };
};

export type RequirementsType = {
  requirement: string;
  comment: string;
  fulfillment: number;
  _id?: string;
  status: number;
};

export type SoftwareDetailsDataType = {
  formDetails: {
    bbDetails: BBDetailsType;
    deploymentCompliance: {
      deploymentCompliance: string;
      deploymentInstructions: string;
    };
  }[];
};

// All types used in Table.tsx and the data connected to it
export type DataType = {
  rows: DataRow[];
};

export type TableProps = {
  data: DataType | Record<string, never>;
  hasVerticalBorders?: boolean;
};

export type CellValue = {
  value: string | number | boolean;
};

export type CellValues = {
  values: CellValue[];
};

export type Cell =
  | CellValue
  | { values: CellValue[] }
  | { values: CellValues[] };

export type DataRow = {
  cell: Cell[];
  subHeader?: string;
};

export type InputOptionsProps = {
  value: unknown;
  label: string;
};

export type InputSingleOptionProps = {
  value: ComplianceRequirementsType;
  label: string;
};

export type POSTSoftwareAttributesType = {
  success: boolean;
  details: string;
  link: string;
  uniqueId: string;
};

export type PATCHSoftwareAttributesType = {
  success: boolean;
  details: string;
  link: string;
};

export type SubmitDraftResponseType = {
  success: boolean;
  details: string;
  link: string;
};

// Types used in IRSC/IRSC...Table.tsx and the data connected to it
export type IRSCTableType = {
  selectedData: ComplianceRequirementsType;
  setUpdatedData: (data: ComplianceRequirementsType) => void;
  isTableValid: boolean;
  readOnlyView?: boolean;
};
