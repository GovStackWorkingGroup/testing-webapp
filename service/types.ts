export type CompatibilitiesType = {
  id: string;
  buildingBlock: string; // Building block label
  timestamp: number; // Execution time from test [seconds]
  saveTime: number; // Save time in db [miliseconds]
  testsPassed: number;
  testsFailed: number;
  compatibility: number;
};

export type ProductsType = {
  _id: {
    testApp: string; // Label of the product
    testSuite: string; // Test suite label
    sourceBranch: string; // Branch on which tests were executed
  };
  compatibilities: CompatibilitiesType[];
};

export type ProductsListType = ProductsType[];
