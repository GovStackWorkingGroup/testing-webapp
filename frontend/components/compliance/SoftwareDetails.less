@import '../../styles/Variables.less';

.compliance-detail-page-container {
  padding: 16px 25px;
}

.software-attributes-section {
  padding: 40px 10%;
}

.software-attributes-section > p {
  padding-bottom: 16px;
}

.software-attributes-container {
  width: 100%;
  border: 1px solid @stroke;
  border-radius: 8px;
}

.software-attribute {
  display: flex;
  align-items: center;
  border-bottom: 1px solid @stroke;
  padding-left: 16px;
}

.software-attributes-container > .software-attribute:last-child {
  border-bottom: none;
}

.software-attribute > div {
  display: flex;
  align-items: center;
  min-height: @table-row;
  font-size: 13px;
  font-weight: 700;
  color: @text-secondary;
}

.software-attribute > div:first-child {
  width: 30%;
  background-color: @blue-lightest;
  border-right: 1px solid @stroke;
}

.software-attribute > div:last-child {
  gap: 10px;
}

.software-attribute > div:first-child {
  width: 20%;
  background-color: @blue-lightest;
  border-right: 1px solid @stroke;
}

.software-attribute > div:last-child {
  gap: 10px;
  padding-left: 16px;
}

.software-attribute-logo > div {
  min-height: 90px;
}

.software-attribute-logo > div:last-child {
  display: flex;
  flex-direction: column;
}

.software-attribute-logo > div > img {
  padding-top: 16px;
}

.img-logo {
  max-width: 60px;
  max-height: 60px;
}

.software-attribute > div > a {
  color: @blue-medium;
}
