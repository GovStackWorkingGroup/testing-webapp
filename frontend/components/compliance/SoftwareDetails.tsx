import { useRouter } from 'next/router';
import useTranslations from '../../hooks/useTranslation';

type SoftwareDetailsProps = {
  title: string;
  children: React.ReactNode;
  complianceSection?: boolean;
  softwareVersion?: string;
  customStyles?: string;
  editButton?: boolean;
  redirectToStep?: number;
};

const SoftwareDetails = ({
  title,
  children,
  complianceSection = false,
  softwareVersion,
  customStyles,
  editButton = false,
  redirectToStep,
}: SoftwareDetailsProps) => {
  const { format } = useTranslations();
  const router = useRouter();

  const { draftUUID } = router.query;

  const handlePressEdit = () => {
    if (draftUUID) {
      router.replace({
        query: { draftUUID, formStep: redirectToStep },
      });
    }
  };

  return (
    <div
      className={`software-attributes-section ${
        customStyles ? customStyles : ''
      }`}
    >
      {complianceSection ? (
        <p>
          {title}{' '}
          <span className="bold">Software version {softwareVersion}</span>
        </p>
      ) : (
        <div className="software-attributes-title-with-link">
          <p>{title}</p>
          {editButton && (
            <p
              className="software-attributes-title-edit-link"
              onClick={() => handlePressEdit()}
            >
              {format('app.edit.label')}
            </p>
          )}
        </div>
      )}

      {children}
    </div>
  );
};

export default SoftwareDetails;
