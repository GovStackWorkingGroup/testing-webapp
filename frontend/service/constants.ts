// Urls
export const COMPLIANCE_TESTING_DETAILS_PAGE =
  'requirements/details/';
export const COMPLIANCE_TESTING_FORM = '/requirements/form';
export const COMPLIANCE_TESTING_RESULT_PAGE = '/requirements';
export const API_TESTING_RESULT_PAGE = '/apiTesting';
export const CONFLUENCE_INSTRUCTIONS_LINK =
  'https://govstack-global.atlassian.net/wiki/spaces/GH/pages/376012801/Instructions+for+Software+Requirements+Compliance';

// Data
export const softwareComplianceFormSteps = [
  {
    label: 'app.software_attributes.label',
    step: 1,
  },
  {
    label: 'table.deployment_compliance.label',
    step: 2,
  },
  {
    label: 'app.api_requirement_specification.label',
    step: 3,
  },
  {
    label: 'app.evaluation_summary.label',
    step: 4,
  },
];

// Local storage
export const SOFTWARE_ATTRIBUTES_STORAGE_NAME = 'software_attributes';
export const DEPLOYMENT_COMPLIANCE_STORAGE_NAME = 'deployment_compliance';
export const INTERFACE_COMPLIANCE_STORAGE_NAME = 'interface_compliance';
export const REQUIREMENT_SPEC_STORAGE_NAME =
  'requirement_specification_compliance';
