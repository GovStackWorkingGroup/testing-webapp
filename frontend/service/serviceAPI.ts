import { FormValuesType } from '../components/form/SoftwareAttributesForm';
import {
  ResultTableSortByType,
  SoftwaresTableSortByType,
} from '../components/table/types';
import {
  BuildingBlockTestSummary,
  ComplianceList,
  ComplianceRequirementsType,
  FormUpdatedObject,
  PATCHSoftwareAttributesType,
  POSTSoftwareAttributesType,
  ProductsListType,
  SoftwareDetailsDataType,
  SoftwareDetailsType,
  SoftwareDraftDetailsType,
  SoftwareDraftToUpdateType,
  SubmitDraftResponseType,
  SubmittingFormResponseType,
} from './types';

export const baseUrl = process.env.API_URL;

type Success<T> = { status: true; data: T };
type Failure = { status: false; error: Error };

const handleFieldsToSort = (
  sortBy: ResultTableSortByType | SoftwaresTableSortByType
) =>
  Object.values(sortBy)
    .filter((order) => order.order !== null)
    .map((sortProperty) => `sort.${sortProperty.field}=${sortProperty.order}`)
    .join('&');

const baseFetch = async (url: string): Promise<ProductsListType> => {
  const response = await fetch(url, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    throw new Error(response.statusText);
  }

  return response.json();
};

export const getSoftwaresData = async (
  offset: number,
  sortBy: SoftwaresTableSortByType,
  branch?: string
): Promise<Success<ProductsListType> | Failure> => {
  try {
    const url = new URL(`${baseUrl}/report/`);
    url.searchParams.set('limit', '20');
    url.searchParams.set('offset', offset.toString());
    url.searchParams.set('branch', branch || 'main');

    const sortedParameters = handleFieldsToSort(sortBy);
    for (const param of sortedParameters.split('&')) {
      const [key, value] = param.split('=');
      url.searchParams.set(key, value);
    }

    const actualData = await baseFetch(url.toString());

    return { data: actualData, status: true };
  } catch (error) {
    return {
      status: false,
      error: error as Error,
    };
  }
};

export const getSoftwareListCount = async (
  branch?: string
): Promise<Success<number> | Failure> => {
  try {
    const url = new URL(`${baseUrl}/report/count`);

    // Assuming branch is a query parameter. If not, this can be removed.
    const params = new URLSearchParams({ branch: branch || 'main' });
    url.search = params.toString();

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      throw new Error(response.statusText);
    }

    const result = await response.json();

    return {
      status: true,
      data: result.count,
    };
  } catch (error) {
    return {
      status: false,
      error: error as Error,
    };
  }
};

export const getBuildingBlockTestResults = async (
  buildingBlockId: string,
  sortBy: ResultTableSortByType
) => {
  const sortedParameters = handleFieldsToSort(sortBy);

  return await fetch(
    `${baseUrl}/report/${buildingBlockId}?${sortedParameters}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<BuildingBlockTestSummary>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getComplianceList = async (offset: number, limit: number) => {

  const accessToken = sessionStorage.getItem('accessToken');

  return await fetch(
    `${baseUrl}/compliance/list?offset=${offset}&limit=${limit}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
        ...(accessToken ? { 'Authorization': `Bearer ${accessToken}` } : {}),
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<ComplianceList>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getSoftwareDetails = async (softwareName: string) => {
  return await fetch(`${baseUrl}/compliance/${softwareName}/detail`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SoftwareDetailsType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const saveSoftwareDraft = async (software: FormValuesType) => {
  const formData = new FormData();
  formData.append('softwareName', software.softwareName.value);
  formData.append('logo', software.softwareLogo.value as File);
  formData.append('website', software.softwareWebsite.value);
  formData.append('documentation', software.softwareDocumentation.value);
  formData.append('description', software.toolDescription.value);
  formData.append('email', software.email.value);
  formData.append('version', software.softwareVersion.value);
  formData.append('compliance[0][version]', software.softwareVersion.value);

  return await fetch(`${baseUrl}/compliance/drafts`, {
    method: 'post',
    body: formData,
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<POSTSoftwareAttributesType>>((response) => {
      return { data: response, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getDraftDetails = async (draftUUID: string) => {
  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SoftwareDraftDetailsType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getComplianceRequirements = async () => {
  return await fetch(`${baseUrl}/compliance/requirements/`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<ComplianceRequirementsType[]>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const updateDraftDetailsStepOne = async (
  draftUUID: string,
  data: FormValuesType
) => {
  const formData = new FormData();

  formData.append('softwareName', data.softwareName.value);
  formData.append('logo', data.softwareLogo.value as File);
  formData.append('website', data.softwareWebsite.value);
  formData.append('documentation', data.softwareDocumentation.value);
  formData.append('description', data.toolDescription.value);
  formData.append('email', data.email.value);
  formData.append('compliance[0][version]', data.softwareVersion.value);

  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'PATCH',
    body: formData,
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<POSTSoftwareAttributesType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const updateDraftDetailsStepTwo = async (
  draftUUID: string,
  data: SoftwareDraftToUpdateType
) => {
  const formData = new FormData();

  if (
    data.deploymentCompliance?.documentation ||
    data.deploymentCompliance?.documentation === ''
  ) {
    if (data.deploymentCompliance?.documentation instanceof File) {
      formData.append(
        'deploymentCompliance[documentation]',
        data.deploymentCompliance.documentation
      );
    } else {
      formData.append(
        'deploymentCompliance[documentation]',
        data.deploymentCompliance?.documentation
      );
    }
  }

  if (
    data.deploymentCompliance?.deploymentInstructions ||
    data.deploymentCompliance?.deploymentInstructions === ''
  ) {
    if (data.deploymentCompliance?.deploymentInstructions instanceof File) {
      formData.append(
        'deploymentCompliance[deploymentInstructions]',
        data.deploymentCompliance?.deploymentInstructions
      );
    } else {
      formData.append(
        'deploymentCompliance[deploymentInstructions]',
        data.deploymentCompliance?.deploymentInstructions || ''
      );
    }
  }

  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'PATCH',
    body: formData,
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<PATCHSoftwareAttributesType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const updateDraftDetailsStepThree = async (
  draftUUID: string,
  data: ComplianceRequirementsType[],
  softwareVersion: string
) => {
  const transformedData = data.map((item) => {
    const {
      bbKey,
      bbName,
      bbVersion,
      dateOfSave,
      requirements,
      interfaceCompliance,
    } = item;
    const test = {
      bbSpecification: bbName,
      bbVersion,
      dateOfSave,
      requirementSpecificationCompliance: {
        crossCuttingRequirements: requirements.crossCutting.map(
          (crossCuttingItem) => ({
            requirement: crossCuttingItem.requirement,
            comment: crossCuttingItem.comment,
            fulfillment: crossCuttingItem.fulfillment,
            _id: crossCuttingItem._id,
          })
        ),
        functionalRequirements: requirements.functional.map(
          (functionalItem) => ({
            requirement: functionalItem.requirement,
            comment: functionalItem.comment,
            fulfillment: functionalItem.fulfillment,
            _id: functionalItem._id,
          })
        ),
      },
      interfaceCompliance: {
        testHarnessResult: interfaceCompliance?.testHarnessResult || '',
        requirements: interfaceCompliance?.requirements || [],
      },
    };

    return {
      [bbKey]: test,
    };
  });

  const transformedDataObject = transformedData.reduce((acc, item) => {
    const bbKey = Object.keys(item)[0];

    if (bbKey) {
      acc[bbKey] = item[bbKey];
    }

    return acc;
  }, {});

  const payload = {
    compliance: {
      version: softwareVersion,
      bbDetails: transformedDataObject,
    },
  };

  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<PATCHSoftwareAttributesType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const submitDraft = async (uniqueId: string) => {
  const payload = { uniqueId };

  return await fetch(`${baseUrl}/compliance/drafts/submit`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SubmitDraftResponseType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getSoftwareDetailsReport = async (id: string) => {
  return await fetch(`${baseUrl}/compliance/forms/${id}`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SoftwareDetailsDataType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const handleReviewSoftwareForm = async (
  id: string,
  data: FormUpdatedObject,
  type: 'update' | 'accept' | 'reject'
) => {
  const accessToken = sessionStorage.getItem('accessToken');

  return await fetch(`${baseUrl}/compliance/forms/${id}/${type}`, {
    method: 'post',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json',
      ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SubmittingFormResponseType>>((response) => {
      return { data: response, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

// This endpoint do not connect to project BE
export const checkIfImageUrlExists = async (url: string): Promise<boolean> => {
  try {
    const response = await fetch(url, { method: 'HEAD' });

    return response.status === 200;
  } catch (error) {
    return false;
  }
};

// Fetch file
export const fetchFileDetails = async (file: string) => {
  const filePath = `${baseUrl}/${file}`;
  try {
    const response = await fetch(filePath, {
      method: 'GET',
      mode: 'cors',
      cache: 'no-cache',
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    const contentType = response.headers.get('content-type') || '';
    const blob = await response.blob();

    // Create a File object from the Blob
    const url = new URL(filePath);

    const pathname = url.pathname;

    const parts = pathname.split('/');

    const fileName = parts[parts.length - 1]; // Provide the desired file name
    const file = new File([blob], fileName, { type: contentType });

    return file;
  } catch (error) {
    return undefined;
  }
};
