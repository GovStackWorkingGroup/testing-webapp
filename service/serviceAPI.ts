import {
  ResultTableSortByType,
  SoftwaresTableSortByType,
} from '../components/table/types';
import { BuildingBlockTestSummary, ProductsListType } from './types';

const baseUrl = 'https://api.testing.govstack.global';

type Success<T> = { status: true; data: T };
type Failure = { status: false; error: Error };

const handleFieldsToSort = (
  sortBy: ResultTableSortByType | SoftwaresTableSortByType
) =>
  Object.values(sortBy)
    .filter((order) => order.order !== null)
    .map((sortProperty) => `sort.${sortProperty.field}=${sortProperty.order}`)
    .join('&');

export const getSoftwaresData = async (
  offset: number,
  sortBy: SoftwaresTableSortByType
) => {
  const sortedParameters = handleFieldsToSort(sortBy);

  return await fetch(
    `${baseUrl}/report/?limit=20&offset=${offset}&${sortedParameters}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return response.json();
    })
    .then<Success<ProductsListType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getSoftwareListCount = async () => {
  return await fetch(`${baseUrl}/report/count`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then(async (response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return response.json();
    })
    .then<Success<number>>((actualCount) => {
      return {
        status: true,
        data: actualCount.count,
      };
    })
    .catch<Failure>((error) => {
      return {
        status: false,
        error,
      };
    });
};

export const getBuildingBlockTestResults = async (
  buildingBlockId: string,
  sortBy: ResultTableSortByType
) => {
  const sortedParameters = handleFieldsToSort(sortBy);

  return await fetch(
    `${baseUrl}/report/${buildingBlockId}?${sortedParameters}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return response.json();
    })
    .then<Success<BuildingBlockTestSummary>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};
