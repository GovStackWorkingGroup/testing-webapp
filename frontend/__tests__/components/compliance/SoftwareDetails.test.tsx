import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { toast } from 'react-toastify';
import { useRouter } from 'next/router';
import SoftwareDetails from '../../../components/compliance/SoftwareDetails';
import { handleDeleteSoftwareForm } from '../../../service/serviceAPI';
import { COMPLIANCE_TESTING_RESULT_PAGE } from '../../../service/constants';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));
jest.mock('react-toastify', () => ({
  toast: {
    success: jest.fn(),
    error: jest.fn(),
  },
}));
jest.mock('../../../service/serviceAPI', () => ({
  handleDeleteSoftwareForm: jest.fn(),
}));
jest.mock('../../../hooks/useTranslation', () => ({
  __esModule: true,
  default: () => ({
    format: (key: string) => key,
  }),
}));

describe('SoftwareDetails component', () => {
  const mockRouter = {
    query: { draftUUID: '123', softwareName: 'TestSoftware' },
    replace: jest.fn(),
    push: jest.fn(),
  };

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the component with basic props', () => {
    render(
      <SoftwareDetails title="Test Title">
        <div>Test Content</div>
      </SoftwareDetails>
    );

    expect(screen.getByText('Test Title')).toBeInTheDocument();
    expect(screen.getByText('Test Content')).toBeInTheDocument();
  });

  it('renders edit button when editButton prop is true', () => {
    render(
      <SoftwareDetails title="Test Title" editButton={true} redirectToStep={2}>
        <div>Test Content</div>
      </SoftwareDetails>
    );

    const editButton = screen.getByText('app.edit.label');
    expect(editButton).toBeInTheDocument();

    fireEvent.click(editButton);
    expect(mockRouter.replace).toHaveBeenCalledWith({
      query: { draftUUID: '123', formStep: 2 },
    });
  });

  it('renders compliance section with delete and view report buttons', () => {
    render(
      <SoftwareDetails
        title="Test Title"
        complianceSection={true}
        softwareVersion="1.0"
        softwareId="456"
        viewReportDetails={true}
      >
        <div>Test Content</div>
      </SoftwareDetails>
    );

    expect(screen.getByText('form.software_delete.label')).toBeInTheDocument();
    expect(screen.getByText('app.view_report_details.label')).toBeInTheDocument();
    expect(screen.getByText('table.software_name.label 1.0')).toBeInTheDocument();
  });

  it('handles software deletion', async () => {
    (handleDeleteSoftwareForm as jest.Mock).mockResolvedValue({ status: true });

    render(
      <SoftwareDetails
        title="Test Title"
        complianceSection={true}
        softwareId="456"
      >
        <div>Test Content</div>
      </SoftwareDetails>
    );

    fireEvent.click(screen.getByText('form.software_delete.label'));
    expect(screen.getByText('software_details.delete_confirm_modal.title')).toBeInTheDocument();

    fireEvent.click(screen.getByText('software_details.delete_confirm_modal.delete'));

    await waitFor(() => {
      expect(handleDeleteSoftwareForm).toHaveBeenCalledWith('456');
      expect(toast.success).toHaveBeenCalledWith('form.form_deleted_success.message', expect.any(Object));
      expect(mockRouter.push).toHaveBeenCalledWith(COMPLIANCE_TESTING_RESULT_PAGE);
    });
  });

  it('handles software deletion error', async () => {
    (handleDeleteSoftwareForm as jest.Mock).mockResolvedValue({ status: false });

    render(
      <SoftwareDetails
        title="Test Title"
        complianceSection={true}
        softwareId="456"
      >
        <div>Test Content</div>
      </SoftwareDetails>
    );

    fireEvent.click(screen.getByText('form.software_delete.label'));
    fireEvent.click(screen.getByText('software_details.delete_confirm_modal.delete'));

    await waitFor(() => {
      expect(handleDeleteSoftwareForm).toHaveBeenCalledWith('456');
      expect(toast.error).toHaveBeenCalledWith('form.form_deleted_error.message', expect.any(Object));
    });
  });

  it('closes delete confirmation modal when cancel is clicked', () => {
    render(
      <SoftwareDetails
        title="Test Title"
        complianceSection={true}
        softwareId="456"
      >
        <div>Test Content</div>
      </SoftwareDetails>
    );

    fireEvent.click(screen.getByText('form.software_delete.label'));
    expect(screen.getByText('software_details.delete_confirm_modal.title')).toBeInTheDocument();

    fireEvent.click(screen.getByText('software_details.delete_confirm_modal.cancel'));
    expect(screen.queryByText('software_details.delete_confirm_modal.title')).not.toBeInTheDocument();
  });
});
