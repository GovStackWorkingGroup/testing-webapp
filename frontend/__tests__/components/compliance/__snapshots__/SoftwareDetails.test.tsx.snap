// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for SoftwareDetails component: should render the component and match snapshot 1`] = `
<div>
  <div
    class="software-attributes-section"
  >
    <p>
      Test software
    </p>
    <div>
      Test
    </div>
  </div>
</div>
`;

exports[`Unit tests for SoftwareDetails component: should render the component and match snapshot when complianceSection is true 1`] = `
<div>
  <div
    class="software-attributes-section"
  >
    <p>
      Test software
       
      <span
        class="bold"
      >
        Software version 
        2.0
      </span>
    </p>
    <div>
      Test
    </div>
  </div>
</div>
`;
